README.md
# lemme-know-bot

[![Build Status][gh-actions-badge]][gh-actions] [![Clojars Project][clojars-badge]][clojars] [![Clojure Docs][cljdoc-badge]][cljdoc-link] [![Clojure version][clojure-v]](project.clj)

A Telegram Bot that mentions you when your specified text in a chat matches.

## Using the Code

Using the lemme-know-bot code in your own project.

To build the service JAR or use the Docker container instead, go to the next section.

See the `lemme-know-bot` [Clojars page][clojars] for details on adding this library to projects for Leiningen, Boot, and CLI/deps.edn.

### Require the Library

In the REPL

```clojure
(require '[lemme-know-bot.core :as lemme])
```

In your application

```clojure
(ns my-app.core
  (:require [lemme-know-bot.core :as lemme]))
```

## Configuration

Supported environment variables and whether they are required to be provided.

| Variable     | Default                            | Description                                            | Required? |
| ------------ | ---------------------------------- | ------------------------------------------------------ | --------- |
| BOT_TOKEN    | none                               | Token for bot to authenticate to the Telegram servers. | Yes       |
| LKB_SEARCHES | "/tmp/lemme-know-bot-searches.edn" | File for saving search state to.                       | No        |
| LKB_SLEEP    | 10000                              | Sleep time in ms between long polls.                   | No        |
| LKB_TIMEOUT  | 10                                 | Timeout in seconds to wait while long polling.         | No        |

## Usage

Building the Java Jar and running it or the Docker container.

### Pre-Reqs

Before building the JAR or running the docker container:

- Create a Telegram bot using the botfather.
  - Configure the bot to give it access to group chat messages. (group privacy disabled)

Proceed with either the [Docker](#run-the-app---docker) or [Jar](#run-the-app---java-jar) instructions.

### Run the App - Docker

- Setup the volume for saving the search state

```bash
# Docker volume name (docker volume ls)
DOCKER_VOL=lemme-know-bot

docker volume create ${DOCKER_VOL}

# Container path for the search state (matches the Dockerfile)
CONTAINER_SAVE_DIR=/usr/src/app/state
```

- Run the container

```bash
docker run \
--detach \
--name lemme-know-bot \
--env BOT_TOKEN="MY-TOKEN-HERE" \
--env LKB_SEARCHES=${CONTAINER_SAVE_DIR}/lemme-know-bot-searches.edn \
--volume ${DOCKER_VOL}:${CONTAINER_SAVE_DIR} \
wdhowe/lemme-know-bot
```

### Pre-Reqs - Java Jar

Pre-reqs for building the Java Jar.

- Install leiningen.
- Clone this project.
- Build the jar

```bash
lein uberjar
```

### Run the App - Java Jar

- Make your bot token available in the environment
  - Other env vars should be exported at this step if you would like to change any default settings.
  - See [Configuration](#configuration) for details.

```bash
export BOT_TOKEN="MY-TOKEN-HERE"
```

- Start the service

```bash
java -jar lemme-know-bot-VERSION-standalone.jar
```

### Telegram Group Setup

Once the service is running from one of the above methods:

- Add your bot to a Telegram group chat.
- Start chatting with your bot in the Telegram group chat.
  - Send '/help' to get started.

## License

Copyright © 2020 Bill Howe

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
<http://www.eclipse.org/legal/epl-2.0>.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at <https://www.gnu.org/software/classpath/license.html>.

----

<!-- Named page links below: /-->

[gh-actions-badge]: https://github.com/wdhowe/lemme-know-bot/workflows/ci%2Fcd/badge.svg
[gh-actions]: https://github.com/wdhowe/lemme-know-bot/actions
[cljdoc-badge]: https://cljdoc.org/badge/lemme-know-bot/lemme-know-bot
[cljdoc-link]: https://cljdoc.org/d/lemme-know-bot/lemme-know-bot/CURRENT
[clojure-v]: https://img.shields.io/badge/clojure-1.10.3-blue.svg
[clojars]: https://clojars.org/lemme-know-bot
[clojars-badge]: https://img.shields.io/clojars/v/lemme-know-bot.svg


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]

- Nothing yet.

## [0.4.0] - 2021-07-24

### 0.4.0 Changed

- Protect get-updates from exiting the bot with any exceptions, log the error instead.
- Send messages on a futures thread to continue processing.
- Updated project dependency versions.

## [0.3.1] - 2020-11-29

### 0.3.1 Changed

- Dependency bump telegrambot-lib 0.3.0 to 0.3.1.

## [0.3.0] - 2020-11-27

### 0.3.0 Changed

- Search lists are now chat+username specific (no longer shared across chats) ([issue-2](https://github.com/wdhowe/lemme-know-bot/issues/2))

## [0.2.1] - 2020-11-26

### 0.2.1 Changed

- Bumped openjdk version in Dockerfile from 8 to 11.

## [0.2.0] - 2020-11-25

### 0.2.0 Added

- Dockerfile for building the service in a container.
- Persistent searches ([issue-1](https://github.com/wdhowe/lemme-know-bot/issues/1))
- Catch HTTP exceptions ([issue-4](https://github.com/wdhowe/lemme-know-bot/issues/4))
- Clean duplicate searches ([issue-5](https://github.com/wdhowe/lemme-know-bot/issues/5))

## 0.1.0 - 2020-11-23

- Initial project release.

---

[Unreleased]: https://github.com/wdhowe/lemme-know-bot/compare/0.4.0...HEAD
[0.4.0]: https://github.com/wdhowe/lemme-know-bot/compare/0.3.1...0.4.0
[0.3.1]: https://github.com/wdhowe/lemme-know-bot/compare/0.3.0...0.3.1
[0.3.0]: https://github.com/wdhowe/lemme-know-bot/compare/0.2.1...0.3.0
[0.2.1]: https://github.com/wdhowe/lemme-know-bot/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/wdhowe/lemme-know-bot/compare/0.1.0...0.2.0

[comment]: # (Types of changes)
[comment]: # ('Added' for new features.)
[comment]: # ('Changed' for changes in existing functionality.)
[comment]: # ('Deprecated' for soon-to-be removed features.)
[comment]: # ('Removed' for now removed features.)
[comment]: # ('Fixed' for any bug fixes.)
[comment]: # ('Security' in case of vulnerabilities.)


test/lemme_know_bot/core_test.clj
(ns lemme-know-bot.core-test
  (:require [clojure.test :refer [deftest is testing]]
            [lemme-know-bot.core :as lemme]))

(deftest create-bot-test
  (testing "Test creating a bot."
    (let [mybot (lemme/create-bot "123")]
      (is (:bot-token mybot) "123"))))


test/lemme_know_bot/fields_test.clj
(ns lemme-know-bot.fields-test
  (:require [clojure.test :refer [deftest is testing]]
            [lemme-know-bot.fields :as fields]))

(def test-group-msg
  "A sample group message with text contents."
  {:update_id 970209698
   :message
   {:message_id 19
    :from {:id 123, :is_bot false, :first_name "Bill", :last_name "Howe", :username "myuser"}
    :chat {:id -456, :title "MyGroup", :type "group", :all_members_are_administrators true}
    :date 1606101165
    :text "/start"}})

(def test-group-system
  "A sample group system message with no text contents."
  {:update_id 862905780
   :message
   {:message_id 161
    :from {:id 123, :is_bot false, :first_name "Bill", :last_name "Howe", :username "myuser", :language_code "en"}
    :chat {:id -456, :title "MyGroup", :type "group", :all_members_are_administrators true}
    :date 1606020445
    :group_chat_created true}})

(def test-updates-msg
  "Sample data from a get-updates call to the Telegram API.
   Group chat example."
  {:ok true
   :result [{:update_id 862905848
             :message {:message_id 286
                       :from {:id 123, :is_bot false
                              :first_name "Bill", :last_name "Howe"
                              :username "myuser", :language_code "en"}
                       :chat {:id -456, :title "MyGroup"
                              :type "group", :all_members_are_administrators true}
                       :date 1606085114, :text "hello"}}]})

(deftest chat-id-test
  (testing "Check the chat id in messages."
    (is (= -456 (fields/chat-id test-group-msg)))))

(deftest chat-results-test
  (testing "Check the get-updates response object for results."
    (let [resp (fields/chat-results test-updates-msg)]
      (is (= 1 (count resp)))
      (is (= "hello" (-> resp
                         (first)
                         :message
                         :text))))))

(deftest from-user-test
  (testing "Check for the from username in the message."
    (is (= "myuser" (fields/from-user test-group-msg)))))

(deftest text-test
  (testing "Check for the text in a test message."
    (is (= "/start" (fields/text test-group-msg)))
    (is (= nil (fields/text test-group-system)))))

(deftest update-id-test
  (testing "Check for the update_id in a test message."
    (is (= 970209698 (fields/update-id test-group-msg)))))


test/lemme_know_bot/notify_test.clj
(ns lemme-know-bot.notify-test
  (:require [clojure.test :refer [are deftest is testing]]
            [lemme-know-bot.notify :as notify]))

(deftest add-search!-test
  (testing "Test adding searches."
    (let [;test chat-id add by integer
          one-search (notify/add-search! 123 "yoda" "hello")
          ;add the rest of the chat-ids by string
          two-searches (notify/add-search! "123" "yoda" "hi there")
          three-searches (notify/add-search! "123" "luke" "ahoy")
          dupe-search (notify/add-search! "123" "luke" "ahoy")]

      ;; check text after adding 2 searches
      (is (= "hello" (:text (first one-search))))
      (is (= "hi there" (:text (second two-searches))))

      ;; count after 3rd search
      (is (= 3 (count three-searches)))

      ;; count after a duplicate
      (is (= 4 (count dupe-search))))))

(deftest clean-searches!-test
  (testing "Test cleaning up duplicates."
    (let [clean-search (notify/clean-searches!)]
    ;; count after removing duplicates
      (is (= 3 (count clean-search))))))

(deftest list-searches-test
  (testing "Test listing searches."
    (is (= 1 (count (notify/list-searches "123" "luke"))))))

(deftest load-searches!-test
  (testing "Testing load-searches! with different data structures."
    (are [func result] (= func result)

      ;; count is 5 due to previous test's manipulation of the atom
      (count (notify/load-searches! [{:chat-id "123" :user "user1", :text "hello"}
                                     {:chat-id "123" :user "user1", :text "hi there"}])) 5

      (count (notify/load-searches! {:user "user1", :text "invalid"})) 0

      (count (notify/load-searches! "invalid")) 0)))

(deftest remove-search!-test
  (testing "Test removing a search item."
    (let [rm-search (notify/remove-search! "123" "yoda" "hi there")]
      ;; count after removing another search
      (is (= 4 (count rm-search))))))


project.clj
(defproject lemme-know-bot "0.4.0"
  :description "A Telegram Bot that mentions you when the specified text in a chat matches."
  :url "https://github.com/wdhowe/lemme-know-bot"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[com.taoensso/timbre "5.1.2"]
                 [environ "1.2.0"]
                 [org.clojure/clojure "1.10.3"]
                 [telegrambot-lib "1.0.0"]
                 [cheshire "5.10.1"]]
  :main ^:skip-aot lemme-know-bot.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}
             :dev [:project/dev :profiles/dev]
             :test [:project/test :profiles/test]
             ;; only edit :profiles/* in profiles.clj
             :profiles/dev {}
             :profiles/test {}
             :project/dev {:env {:log-level "debug"}
                           :plugins [[lein-environ "1.1.0"]]}
             :project/test {:env {:log-level "info"}
                            :plugins [[lein-environ "1.1.0"]]}})


src/lemme_know_bot/core.clj
(ns lemme-know-bot.core
  "Core service and application logic."
  (:gen-class)
  (:require [clojure.string :as string]
            [lemme-know-bot.config :as cfg]
            [lemme-know-bot.fields :as fields]
            [lemme-know-bot.handlers :as handlers]
            [lemme-know-bot.notify :as notify]
            [lemme-know-bot.status :as status]
            [taoensso.timbre :as log]
            [telegrambot-lib.core :as tbot]))

;; Commands for use with the botfather
;; add - Add a new keyword or phrase to watch for in this chat. (parameter required)
;; list - List all watched for keywords or phrases. (no parameter)
;; remove - Remove a watched for keyword or phrase. (parameter required)
;; help - Show a help message with available commands.
;; settings - Show the bot configuration.

(defn create-bot
  "Create a Telegram bot instance."
  ([]
   (tbot/create))
  ([token]
   (tbot/create token)))

(comment (create-bot)
         (create-bot "123"))

(defn poll-updates
  "Long poll for recent chat messages from Telegram."
  ([bot]
   (poll-updates bot nil))

  ([bot offset]
   (try
     (tbot/get-updates bot {:offset offset
                            :timeout (:timeout cfg/config)})

     (catch Exception e
       (log/error "tbot/get-updates exception:" e)))))

(defn handle-msg
  "Check the message text for command or string matches and handle the
   message appropriately."
  [bot msg]
  (log/debug "msg received:" msg)

  (let [msg-text (fields/text msg)]

    (when (not (= nil msg-text))
      (cond
        (string/starts-with? msg-text "/start") (handlers/help bot msg)
        (string/starts-with? msg-text "/help") (handlers/help bot msg)
        (string/starts-with? msg-text "/settings") (handlers/settings bot msg)
        (string/starts-with? msg-text "/list") (handlers/list-search bot msg)
        (string/starts-with? msg-text "/add") (handlers/add-search bot msg)
        (string/starts-with? msg-text "/remove") (handlers/remove-search bot msg)
        :else (handlers/search bot msg)))))

(defn app
  "Retrieve and process chat messages."
  [bot]
  (log/info "lemme-know-bot service started.")

  (loop []
    (log/debug "checking for chat updates.")
    (let [updates (poll-updates bot @status/update-id)
          messages (fields/chat-results updates)]

      ;; Check all messages, if any, for commands/keywords.
      (doseq [msg messages]
        (handle-msg bot msg)
        ;; Increment the next update-id to process.
        (-> msg
            (fields/update-id)
            (inc)
            (status/set-id!)))

      ;; Wait a while before checking for updates again.
      (Thread/sleep (:sleep cfg/config)))
    (recur)))

(defn edn->searches
  "Attempt to load an external edn file into the searches atom."
  []
  (let [loaded-search (cfg/load-edn (:searches cfg/config))]
    (when (some? loaded-search)
      (log/info "loading previous searches.")
      (notify/load-searches! loaded-search))))

(defn shutdown-service
  "Shutdown the service cleanly."
  []
  (shutdown-agents)

  (when (seq @notify/searches)
    (log/info "saving searches list to:" (:searches cfg/config))
    (cfg/save-file (into [] @notify/searches) (:searches cfg/config)))

  (log/info "lemme-know-bot service exited."))

(defn -main
  "Create the Telegram bot and run the application."
  []
  (log/info "starting lemme-know-bot service.")

  (edn->searches)

  (.addShutdownHook (Runtime/getRuntime)
                    (Thread. ^Runnable shutdown-service))

  (let [lemme-bot (create-bot)]

    (if (some? (:bot-token lemme-bot))
      (app lemme-bot)
      (log/error "required bot-token not found! exiting."))))


src/lemme_know_bot/config.clj
(ns lemme-know-bot.config
  (:gen-class)
  (:require [clojure.edn :as edn]
            [environ.core :as environ]
            [clojure.java.io :as io]
            [taoensso.timbre :as log]))

(def config
  {:timeout (or (environ/env :lkb-timeout) 10) ; timeout in seconds to wait for long polling
   :sleep (or (environ/env :lkb-sleep) 10000); sleep in milliseconds between polls
   :searches (or (environ/env :lkb-searches) "/tmp/lemme-know-bot-searches.edn")})

(comment
  (println "config is:" config))

(defn load-edn
  "Load edn files from an io/reader `source` (filename or io/resource).
   Orig fn: https://clojuredocs.org/clojure.edn/read#example-5a68f384e4b09621d9f53a79"
  [source]
  (try
    (with-open [r (io/reader source)]
      (edn/read (java.io.PushbackReader. r)))

    (catch java.io.IOException e
      (log/errorf "couldn't open '%s': %s" source (.getMessage e)))
    (catch RuntimeException e
      (log/errorf "error parsing edn file '%s': %s" source (.getMessage e)))))

(defn save-file
  "Save a data structure to a file."
  [content dest]
  (try
    (spit dest content)

    (catch java.io.IOException e
      (log/errorf "couldn't write to '%s': %s" dest (.getMessage e)))))


src/lemme_know_bot/notify.clj
(ns lemme-know-bot.notify
  "Functions to add, list, remove, track the notify search list."
  (:gen-class)
  (:require [taoensso.timbre :as log]))

(defonce searches (atom []))

(defprotocol search-data
  "Only load supported data types."
  (load-searches! [content]
    "Load a searches data structure at once, merging content with an existing."))

(extend-protocol search-data
  clojure.lang.PersistentVector
  (load-searches!
    [content]
    (swap! searches into content))

  Object
  (load-searches!
    [content]
    (log/warn "not loading searches. data is not a supported type.")))

(comment
  (load-searches! [{:chat-id "123", :user "user1", :text "my search1"}])
  (load-searches! [{:chat-id "123", :user "user2", :text "my search2"}])
  (load-searches! {:user "user1" :text "invalid"})
  (load-searches! "invalid"))

(defn add-search!
  "Add a search to the searches atom vector."
  [chat-id user search]
  (let [chat-str (str chat-id)]
    (swap! searches conj {:chat-id chat-str
                          :user user
                          :text search})))

(comment
  (add-search! 123 "yoda" "yo")
  (add-search! "123" "yoda" "ahoy")
  (add-search! "123" "luke" "ahoy")
  (add-search! "456" "luke" "cheeseburger")
  (println @searches))

(defn clean-searches!
  "Cleanup searches by ensuring only distinct user+text exists."
  []
  (swap! searches distinct))

(comment
  (clean-searches!))

(defn list-searches
  "List the tracked searches given a user.
   Returns a lazy sequence of search words."
  [chat-id user]
  (let [chat-str (str chat-id)]
    (->> @searches
         (filter (fn [entry] (and (= (:chat-id entry) chat-str)
                                  (= (:user entry) user))))
         (keep :text))))

(comment
  (println @searches)
  (list-searches "123" "yoda"))

(defn remove-search!
  "Remove a search in the searches atom vector."
  [chat-id user search]
  (let [chat-str (str chat-id)]
    (swap! searches #(remove
                      (fn [entry] (= entry {:chat-id chat-str
                                            :user user
                                            :text search}))
                      %))))

(comment
  (println @searches)
  (remove-search! 123 "yoda" "yo"))


src/lemme_know_bot/handlers.clj
(ns lemme-know-bot.handlers
  "Functions that take action on various message text contents."
  (:gen-class)
  (:require [clojure.string :as string]
            [lemme-know-bot.config :as cfg]
            [lemme-know-bot.fields :as fields]
            [lemme-know-bot.notify :as notify]
            [taoensso.timbre :as log]
            [telegrambot-lib.core :as tbot]))

(defn send-msg
  "Send a message to the chat."
  [bot chat-id text]
  (try
    (tbot/send-message bot chat-id text)

    (catch Exception e
      (log/error "tbot/send-message exception:" e))))

(defn help
  "Send info about this bot and available commands."
  [bot msg]
  (log/debug "/help or /start command received.")

  (let [chat-id (fields/chat-id msg)]
    (future (send-msg bot chat-id
                      (str "-- Lemme-Know-Bot: Help/Getting Started --\n"
                           "This bot can watch the current chat for keywords/phrases "
                           "and mention you when one has been seen.\n\n"
                           "The following commands are supported:\n"
                           "/add keyword - add a new keyword to watch for in this chat.\n"
                           "/list - list all the watched keywords.\n"
                           "/remove keyword - remove a watched keyword.\n"
                           "/help - show this help message with available commands.\n"
                           "/settings - show the bot configuration.")))))

(defn settings
  "Implements the /settings command."
  [bot msg]
  (log/debug "/settings command received.")

  (let [chat-id (fields/chat-id msg)
        cfg-timeout (:timeout cfg/config)
        cfg-sleep (:sleep cfg/config)]
    (future (send-msg bot chat-id
                      (str "-- Lemme-Know-Bot: Settings --\n"
                           "Timeout (secs) to wait during a long poll: " cfg-timeout "\n"
                           "Sleep (ms) between polls: " cfg-sleep)))))

(defn list-search
  "List the current searches setup for the user."
  [bot msg]
  (log/debug "/list command received.")

  (let [chat-id (fields/chat-id msg)
        username (fields/from-user msg)]
    (future (send-msg bot chat-id
                      (str username ", your searches are: "
                           (into [] (notify/list-searches chat-id username)))))))

(defn extract-keyword
  "Extract only the keyword(s) from a text message.
   Convert to lowercase to avoid case sensitive searches."
  [msg]
  (-> msg
      (fields/text)
      (string/replace-first #"/(\w+)@*(\w+)" "")
      (string/trim)
      (string/lower-case)))

(comment
  (extract-keyword {:message {:text "/add"}}) ; no text
  (extract-keyword {:message {:text "/add "}}) ; spaces only
  (extract-keyword {:message {:text "/add hello"}}) ; text
  (extract-keyword {:message {:text "/add hello there"}}) ; two words
  )

(defn add-search
  "Add a new search for the user."
  [bot msg]
  (log/debug "/add command received.")

  (let [add-text (extract-keyword msg)
        chat-id (fields/chat-id msg)
        username (fields/from-user msg)]

    (if (not (string/blank? add-text))
      ;; not blank - add the text and message back
      (do
        (notify/add-search! chat-id username add-text)
        (notify/clean-searches!)

        (future (send-msg bot chat-id
                          (str username ", search added for: '" add-text "'"))))

      ;; blank text - do not add
      (future (send-msg bot chat-id
                        (str username ", search NOT added. "
                             "Passed parameter text was blank.\n"
                             "Try sending something like: /add cheeseburger"))))))

(defn remove-search
  "Remove a search for the user."
  [bot msg]
  (log/debug "/remove command received.")

  (let [rm-text (extract-keyword msg)
        chat-id (fields/chat-id msg)
        username (fields/from-user msg)]

    (if (not (string/blank? rm-text))
      ;; not blank - remove and message back
      (do
        (notify/remove-search! chat-id username rm-text)

        (future (send-msg bot chat-id
                          (str username ", search removed for: '" rm-text "'"))))

      ;; blank text - do not remove
      (future (send-msg bot chat-id
                        (str username ", search NOT removed. "
                             "Passed parameter text was blank.\n"
                             "Try sending someting like: /remove cheeseburger"))))))

(defn search
  "Search for a match in the text.
   Mention the user if found."
  [bot msg]
  (let [msg-text (string/lower-case (fields/text msg))
        chat-id (fields/chat-id msg)
        chat-entries (filter #(= (str chat-id)
                                 (:chat-id %))
                             @notify/searches)]

    (log/debug "searching for keywords in:" msg-text)

    (future (doseq [entry chat-entries]
              (when (string/includes? msg-text (:text entry))
                (send-msg bot chat-id
                          (str "@" (:user entry)
                               " search string '"
                               (:text entry)
                               "' was found in message:\n"
                               "'" msg-text "'")))))))


src/lemme_know_bot/fields.clj
(ns lemme-know-bot.fields
  "Functions that retrieve message fields."
  (:gen-class))

(defn chat-id
  "Returns the chat-id from a message."
  [msg]
  (-> msg
      :message
      :chat
      :id))

(defn chat-results
  "Get results from the chat updates map.
   Returns a list of maps."
  [msg]
  (-> msg
      :result))

(defn from-user
  "Returns the username of the user that the message is from."
  [msg]
  (-> msg
      :message
      :from
      :username))

(defn text
  "Retrieve the text part of the message."
  [msg]
  (-> msg
      :message
      :text))

(defn update-id
  "Returns the update_id part of the message."
  [msg]
  (-> msg
      :update_id))

(comment
  (def test-group-message
    "A sample group chat message with text."
    {:update_id 970209698
     :message
     {:message_id 19
      :from {:id 123, :is_bot false, :first_name "Bill", :last_name "Howe", :username "myuser"}
      :chat {:id -456, :title "MyGroup", :type "group", :all_members_are_administrators true}
      :date 1606101165
      :text "/start"}})

  (def test-group-system
    "Test a system message in a group chat - no text field."
    {:update_id 862905780
     :message
     {:message_id 161
      :from {:id 123, :is_bot false, :first_name "Bill", :last_name "Howe", :username "myuser", :language_code "en"}
      :chat {:id -456, :title "MyGroup", :type "group", :all_members_are_administrators true}
      :date 1606020445
      :group_chat_created true}})

  (def test-updates-msg
    "Data back from a getUpdates call to the Telegram API."
    {:ok true
     :result [{:update_id 862905848
               :message {:message_id 286
                         :from {:id 123, :is_bot false
                                :first_name "Bill", :last_name "Howe"
                                :username "myuser", :language_code "en"}
                         :chat {:id -456, :title "MyGroup"
                                :type "group", :all_members_are_administrators true}
                         :date 1606085114, :text "hello"}}]})

  (chat-id test-group-message)
  (from-user test-group-message)
  (text test-group-message)
  (update-id test-group-message)

  (chat-id test-group-system)
  (from-user test-group-system)
  (text test-group-system)
  (update-id test-group-system)

  (chat-results test-updates-msg))
